### PR TITLE
Add Init for TF_Filesystem GCS

### PR DESCRIPTION
**Pull request by vnghia**
@mihaimaruseac 
This PR add the `init` function for `filesystem_ops`

### CODE REVIEW

**Review by mihaimaruseac — APPROVED** [review]

**Comment by vnvo2409** [comment]
@mihaimaruseac 
Can you merge this PR so I can continue to work. Thank you

**Comment by mihaimaruseac** [comment]
Working on it. There are internal errors that require manual fixing/investigation.

File: tensorflow/workspace.bzl [hunk 1]
```diff
@@ -1126,6 +1126,39 @@ def tf_repositories(path_prefix = "", tf_repo_name = ""):
...
+    tf_http_archive(
+        name = "com_github_googleapis_google_cloud_cpp",
+        sha256 = "839b2d4dcb36a671734dac6b30ea8c298bbeaafcf7a45ee4a7d7aa5986b16569",
+        strip_prefix = "google-cloud-cpp-1.14.0",
+        patch_file = clean_dep("//third_party:com_github_googleapis_google_cloud_cpp.patch"),
+        urls = [
+            "https://storage.googleapis.com/mirror.tensorflow.org/github.com/googleapis/google-cloud-cpp/archive/v1.14.0.tar.gz",
+            "https://github.com/googleapis/google-cloud-cpp/archive/v1.14.0.tar.gz",
+        ],
+    )
```
Comment (mihaimaruseac) [inline tensorflow/workspace.bzl:1139]: This dependency already exists and that is causing the build to fail. Do we need this one or is the old one ok?

https://github.com/tensorflow/tensorflow/blob/1145cc3855b216f872cb0596a6cf90c86edc9b88/tensorflow/workspace.bzl#L317-L340

File: third_party/com_github_googleapis_google_cloud_cpp.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/google/cloud/storage/BUILD b/google/cloud/storage/BUILD
+index d48b297..a7dd6a6 100644
+--- a/google/cloud/storage/BUILD
++++ b/google/cloud/storage/BUILD
+@@ -76,7 +76,7 @@ cc_library(
+         "//google/cloud:google_cloud_cpp_common",
```
Comment (mihaimaruseac) [inline third_party/com_github_googleapis_google_cloud_cpp.patch:6]: Let's remove this line from the diff. It causes errors internally and is not really needed

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: third_party/com_github_googleapis_google_cloud_cpp.patch (hunk 2 above)
Comment (vnvo2409) [reply third_party/com_github_googleapis_google_cloud_cpp.patch:6]: Ok. so we will keep the google-cloud-cpp 1.14 and remove this line ?

Comment (mihaimaruseac) [reply third_party/com_github_googleapis_google_cloud_cpp.patch:6]: We need to remove this line in the patch if we keep google-cloud-cpp 1.14. But if we keep it we also need to move all users of the other google-cloud-cpp to the new library. Let's make one PR that just handles the changes to `third_party` and `workspace.bzl` so that we can test all downstream projects that they are ok with upgrading to `google-cloud-cpp` 1.14. Then, we rebase this one on top of that one.

**Comment by vnvo2409** [comment]
@mihaimaruseac 
https://github.com/tensorflow/tensorflow/pull/40351
I have made a PR that updates `google-cloud-cpp` to `1.14.0`. It also contains the explanations for your reviews.

**Comment by vnvo2409** [comment]
@mihaimaruseac 
`repo_mapping` seems good. Please take a look. Thank you !

File: tensorflow/workspace.bzl [hunk 3]
```diff
@@ -318,6 +318,9 @@ def tf_repositories(path_prefix = "", tf_repo_name = ""):
         name = "com_github_googlecloudplatform_google_cloud_cpp",
         sha256 = "839b2d4dcb36a671734dac6b30ea8c298bbeaafcf7a45ee4a7d7aa5986b16569",
         strip_prefix = "google-cloud-cpp-1.14.0",
+        repo_mapping = {
```
Comment (mihaimaruseac) [inline tensorflow/workspace.bzl:321]: Let's also split changes to this file into a separate PR.

Reason: we need to do additional changes internally to import this and it is better to have that in a single purpose changelist so it is easy to rollback if needed

**Comment by vnvo2409** [comment]
@mihaimaruseac 
Here is a PR for `repo_mapping` only https://github.com/tensorflow/tensorflow/pull/40389

**Comment by mihaimaruseac** [comment]
I was thinking to move the other 2 packages in `workspace.bzl` to a separate PR. We can file a concurrent PR for that.

**Comment by vnvo2409** [comment]
@mihaimaruseac 
Here is a PR to add 2 packages to `workspace.bzl`

**Comment by mihaimaruseac** [comment]
Now we just need to rebase this back on master and then we should be able to merge. I'll probably still need to do some manual import, but it's going to be easier this time and less risky. Maybe by tomorrow we can merge it

**Comment by vnvo2409** [comment]
@mihaimaruseac I have rebased

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
Needs manual import, working on it right now

**Comment by vnvo2409** [comment]
@mihaimaruseac 
I see. Could you give me some idea how internal import work so I can avoid errors in the next PRs ?

**Comment by mihaimaruseac** [comment]
I don't think it's your fault now. Basically internally some parts of the code have different names, some dependencies are in different location. Everytime you have an `@dep` in a `BUILD` file, internally that one is a normal `//path/to/dep` dependency.

There is a script that does all of these conversions but when new things are added the script also needs to be modified. It can be modified only via doing a manual import and fix. That is the reason I suggested splitting this into several PRs, so that we can localize exactly which places need to be changed to satisfy the script and so that it is easy to revert if we did a mistake/broke some build.

**Comment by vnvo2409** [comment]
I see. Thank you

**Comment by mihaimaruseac** [comment]
The Linux GPU failure is unrelated, we are rolling back the change causing the error there. So far public tests seem to be ok, there are a few more internal tests that are running now but everything looks good

**Comment by vnvo2409** [comment]
After this PR is merged, I will send you a PR for WritableFile
